server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDecider.java

```diff
@@ -0,0 +1,167 @@
+            totalShards = indexMetadata.getNumberOfShards() * indexMetadata.getNumberOfReplicas();
+            nomenclature = "search";
+        }
+
+        assert eligibleNodes.isEmpty() == false;
+        assert totalShards > 0;
+
+        final double idealAllocation = Math.ceil((double) totalShards / eligibleNodes.size());
+        final int threshold = (totalShards + eligibleNodes.size() - 1 + indexBalanceConstraintSettings.getExcessShards()) / eligibleNodes
+            .size();
+        final int currentAllocation = node.numberOfOwningShardsForIndex(index);
+
+        if (currentAllocation >= threshold) {
+            String explanation = Strings.format(
+                "There are [%d] eligible nodes in the [%s] tier for assignment of [%d] shards in index [%s]. Ideally no more than [%.0f] "
+                    + "shard would be assigned per node (the index balance excess shards setting is [%d]). This node is already assigned"
+                    + " [%d] shards of the index.",
+                eligibleNodes.size(),
+                nomenclature,
+                totalShards,
+                index,
+                idealAllocation,
+                indexBalanceConstraintSettings.getExcessShards(),
+                currentAllocation
+            );
+
+            logger.trace(explanation);
+
+            return allocation.decision(Decision.NOT_PREFERRED, NAME, explanation);
+        }
+
+        return allocation.decision(Decision.YES, NAME, "Node index shard allocation is under the threshold.");
+    }
+
+    private void collectEligibleNodes(RoutingAllocation allocation, Set<DiscoveryNode> eligibleNodes, DiscoveryNodeRole role) {
+        for (DiscoveryNode discoveryNode : allocation.nodes()) {
+            if (discoveryNode.getRoles().contains(role) && allocation.metadata().nodeShutdowns().contains(discoveryNode.getId()) == false) {
+                eligibleNodes.add(discoveryNode);
+            }
+        }
+    }
+
+    private void setClusterRequireFilters(Map<String, List<String>> filters) {
+        clusterRequireFilters = DiscoveryNodeFilters.trimTier(DiscoveryNodeFilters.buildFromKeyValues(AND, filters));
+    }
+
+    private void setClusterIncludeFilters(Map<String, List<String>> filters) {
+        clusterIncludeFilters = DiscoveryNodeFilters.trimTier(DiscoveryNodeFilters.buildFromKeyValues(OR, filters));
+    }
+
+    private void setClusterExcludeFilters(Map<String, List<String>> filters) {
+        clusterExcludeFilters = DiscoveryNodeFilters.trimTier(DiscoveryNodeFilters.buildFromKeyValues(OR, filters));
+    }
+
+    private boolean hasNoFilters() {
+        return (clusterExcludeFilters == null || clusterExcludeFilters.hasNoFilters())
+            && (clusterIncludeFilters == null || clusterIncludeFilters.hasNoFilters())
+            && (clusterRequireFilters == null || clusterRequireFilters.hasNoFilters());
+    }
```

Comment:
Do we need to keep the actual filters, or rather just track whether any are present here?

i.e could these be tracked as booleans instead e.g.

```java
clusterIncludeFilters = filters.size() > 0;
```

Doesn't seem like a big deal, but best to keep the state/logic minimal where we can, also might mean we can avoid adding `hasNoFilters` ?